core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts

```diff
@@ -47,29 +48,40 @@ class TestEngineRunner {
   }
 
   /** Run request against the engine. The result will be stored in rendering property. */
-  request(url: string) {
-    const response = {};
+  request(url: string, params?: { httpHeaders?: { [key: string]: string } }) {
+    const response: { [key: string]: string } = {};
+    const headers = new Headers(params?.httpHeaders ?? {});
     const optionsMock = {
-      req: {
+      req: <Partial<Request>>{
         originalUrl: url,
+        headers,
+        get: (header: string): string | null => headers.get(header),
       },
-      res: {
+      res: <Partial<Response>>{
         set: (key: string, value: any) => (response[key] = value),
       },
     };
 
     this.engineInstance(url, optionsMock, (_, html) => {
-      this.renders.push(html);
+      this.renders.push(html ?? '');
       this.responseParams.push(response);
     });
 
     return this;
   }
 }
 
+const getCurrentConcurrency = (
+  engineRunner: TestEngineRunner
+): { currentConcurrency: number } => {
```

Comment:
Due to a lot of assertions on numbers 0,1,2,3 inside one it() and because we check in turns: currentConcurrency as well as renderCallbacksCount, I deliberately wrapped the expected value in a meaningful object. Thanks to that the failed test doesn't tell "expected 3 to be 0", but expected { currentConcurrency: 3} to equal { currentConcurrency: 0 }

I suspect I'm solving the readability issue in a fancy way, but the real problem might be somewhere else (e.g. I perform too many `expect`s in one `it`)...